File: scripts/read-parquet.js
const parquet = require('parquetjs');
const fs = require('fs');
const path = require('path');

async function readParquetFile(filePath) {
  try {
    console.log(`Reading parquet file: ${filePath}`);
    
    // Read the parquet file
    const reader = await parquet.ParquetReader.openFile(filePath);
    const cursor = reader.getCursor();
    
    // Get schema information
    const schema = reader.getSchema();
    console.log('\n📊 Schema Information:');
    console.log('Fields:', schema.fieldList.map(f => f.name).join(', '));
    console.log(`Total rows: ${reader.getRowCount()}\n`);
    
    // Read all rows
    const rows = [];
    let row;
    let count = 0;
    
    while (row = await cursor.next()) {
      rows.push(row);
      count++;
      
      // Show first few rows as preview
      if (count <= 5) {
        console.log(`Row ${count}:`, JSON.stringify(row, null, 2));
      }
    }
    
    await reader.close();
    
    console.log(`\n✅ Successfully read ${rows.length} rows from parquet file\n`);
    
    return { rows, schema };
  } catch (error) {
    console.error('❌ Error reading parquet file:', error);
    throw error;
  }
}

function convertToCSV(rows, outputPath) {
  try {
    if (rows.length === 0) {
      console.log('⚠️  No rows to convert');
      return;
    }
    
    // Get all column names from the first row
    const columns = Object.keys(rows[0]);
    
    // Create CSV header
    const header = columns.join(',');
    
    // Create CSV rows
    const csvRows = rows.map(row => {
      return columns.map(col => {
        const value = row[col];
        // Handle values that might contain commas or quotes
        if (value === null || value === undefined) {
          return '';
        }
        const stringValue = String(value);
        // Escape quotes and wrap in quotes if contains comma or quote
        if (stringValue.includes(',') || stringValue.includes('"') || stringValue.includes('\n')) {
          return `"${stringValue.replace(/"/g, '""')}"`;
        }
        return stringValue;
      }).join(',');
    });
    
    // Combine header and rows
    const csvContent = [header, ...csvRows].join('\n');
    
    // Write to file
    fs.writeFileSync(outputPath, csvContent, 'utf8');
    console.log(`✅ Successfully converted to CSV: ${outputPath}`);
    console.log(`   Total rows: ${rows.length}`);
    console.log(`   Total columns: ${columns.length}`);
  } catch (error) {
    console.error('❌ Error converting to CSV:', error);
    throw error;
  }
}

async function main() {
  const args = process.argv.slice(2);
  const filePath = args[0] || 'instrument_master_2025-12-04 1.parquet';
  const outputPath = args[1] || 'instrument_master_2025-12-04.csv';
  const convertOnly = args.includes('--convert-only');
  const readOnly = args.includes('--read-only');
  
  // Resolve file paths relative to project root
  const projectRoot = path.resolve(__dirname, '..');
  const fullFilePath = path.isAbsolute(filePath) ? filePath : path.join(projectRoot, filePath);
  const fullOutputPath = path.isAbsolute(outputPath) ? outputPath : path.join(projectRoot, outputPath);
  
  if (!fs.existsSync(fullFilePath)) {
    console.error(`❌ File not found: ${fullFilePath}`);
    process.exit(1);
  }
  
  try {
    const { rows, schema } = await readParquetFile(fullFilePath);
    
    if (!readOnly) {
      console.log('\n🔄 Converting to CSV...');
      convertToCSV(rows, fullOutputPath);
    } else {
      console.log('\n📋 Read-only mode: CSV conversion skipped');
    }
    
    console.log('\n✨ Done!');
  } catch (error) {
    console.error('❌ Failed:', error.message);
    process.exit(1);
  }
}

// Run the script
main();

File: scripts/read_parquet.py
#!/usr/bin/env python3
"""
Script to read and convert parquet files to CSV
Usage:
    python scripts/read_parquet.py [input_file] [output_file] [--read-only]
"""

import sys
import os
import pandas as pd
import argparse
from pathlib import Path

def read_parquet_file(file_path):
    """Read a parquet file and return the DataFrame"""
    try:
        print(f"Reading parquet file: {file_path}")
        
        # Read the parquet file
        df = pd.read_parquet(file_path, engine='pyarrow')
        
        print(f"\nFile Information:")
        print(f"   Total rows: {len(df)}")
        print(f"   Total columns: {len(df.columns)}")
        print(f"\nColumn Names:")
        for i, col in enumerate(df.columns, 1):
            print(f"   {i}. {col}")
        
        print(f"\nFirst 5 rows preview:")
        print(df.head().to_string())
        
        print(f"\nData Types:")
        print(df.dtypes.to_string())
        
        print(f"\nSuccessfully read {len(df)} rows from parquet file\n")
        
        return df
    except Exception as error:
        print(f"Error reading parquet file: {error}")
        raise

def convert_to_csv(df, output_path):
    """Convert DataFrame to CSV"""
    try:
        if df.empty:
            print("Warning: No data to convert")
            return
        
        # Convert to CSV
        df.to_csv(output_path, index=False)
        
        print(f"Successfully converted to CSV: {output_path}")
        print(f"   Total rows: {len(df)}")
        print(f"   Total columns: {len(df.columns)}")
    except Exception as error:
        print(f"Error converting to CSV: {error}")
        raise

def main():
    parser = argparse.ArgumentParser(description='Read and convert parquet files to CSV')
    parser.add_argument('input_file', nargs='?', 
                       default='instrument_master_2025-12-04 1.parquet',
                       help='Input parquet file path')
    parser.add_argument('output_file', nargs='?',
                       default='instrument_master_2025-12-04.csv',
                       help='Output CSV file path')
    parser.add_argument('--read-only', action='store_true',
                       help='Only read the file, do not convert to CSV')
    
    args = parser.parse_args()
    
    # Resolve file paths
    project_root = Path(__file__).parent.parent
    input_path = Path(args.input_file) if Path(args.input_file).is_absolute() else project_root / args.input_file
    output_path = Path(args.output_file) if Path(args.output_file).is_absolute() else project_root / args.output_file
    
    if not input_path.exists():
        print(f"File not found: {input_path}")
        sys.exit(1)
    
    try:
        # Read the parquet file
        df = read_parquet_file(str(input_path))
        
        # Convert to CSV if not read-only
        if not args.read_only:
            print("\nConverting to CSV...")
            convert_to_csv(df, str(output_path))
        else:
            print("\nRead-only mode: CSV conversion skipped")
        
        print("\nDone!")
    except Exception as error:
        print(f"Failed: {error}")
        sys.exit(1)

if __name__ == '__main__':
    main()

